### PR TITLE
Type parsing more strongly

### LINKED ISSUE #9: Implement Quote and tests

### PR DESCRIPTION
**Pull request by phoebe-lew**
closes #9 , #13 and refactors types

### CODE REVIEW

File: protocol/src/test/kotlin/models/MessageTest.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package protocol.models
```
Comment (KendallWeihe) [inline protocol/src/test/kotlin/models/MessageTest.kt:1]: Within our tests we have packages...

- `models`
- `protocol`
- `protocol.models` 

We should reconcile

File: protocol/src/main/kotlin/models/Close.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package models
+
+import typeid.TypeID
```
Comment (KendallWeihe) [inline protocol/src/main/kotlin/models/Close.kt:3]: Just as an aside, when we first introduced the [`typeid-js` dependency](https://github.com/jetpack-io/typeid-js) in `tbdex-js` we married ourselves to that particular package, which may or may not have multi-language support. In this case it appears it does have multi-language support. But, just a forewarning on rubber stamping new dependencies.

File: protocol/src/main/kotlin/models/Message.kt [hunk 3]
```diff
@@ -2,88 +2,83 @@ package models
...
-        if (signature != null) {
-            verify()
-        } else {
-            validate()
-        }
-    }
 
-    private fun verify() {
-        validate()
+sealed class Message {
```
Comment (andresuribe87) [inline protocol/src/main/kotlin/models/Message.kt:26]: Should this belong in the web5 repo? Seems related to DWNs, and not specifically to tbdex.

File: protocol/src/main/kotlin/Mapper.kt [hunk 4]
```diff
@@ -1,25 +1,18 @@
...
-    fun writer(): ObjectWriter {
-        return objectMapper.writer()
-    }
-
-    fun <T> reader(type: Class<T>): ObjectReader {
-        return objectMapper.readerFor(type)
-    }
-
+  // TODO make private and add method for calling readTree
+  val objectMapper: ObjectMapper = ObjectMapper()
```
Comment (andresuribe87) [inline protocol/src/main/kotlin/Mapper.kt:10]: Curious why you went with jackson for json serialization vs [kotlinx serialization](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/serialization-guide.md)?

File: protocol/src/main/kotlin/models/Message.kt (hunk 3 above)
Comment (ethan-tbd) [reply protocol/src/main/kotlin/models/Message.kt:26]: @andresuribe87 this is actually a tbdex message as described in the tbdex protocol [here](https://github.com/TBD54566975/tbdex/tree/main/specs/protocol#messages)

File: protocol/src/main/kotlin/models/Quote.kt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+) : Message() {
+
+  companion object {
+    fun create(
+      to: String,
+      from: String,
+      exchangeID: TypeID,
+      data: QuoteData,
+    ): Quote {
+      val id = TypeID(MessageKind.quote.name)
```
Comment (ethan-tbd) [inline protocol/src/main/kotlin/models/Quote.kt:19]: super nit: looks like other message classes just call `TypeID()` inside of the `MessageMetadata` constructor on line 20 :))

File: protocol/src/main/kotlin/models/MessageData.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  val offeringID: TypeID,
+  val payinSubunits: Int,
+  val payinMethod: SelectedPaymentMethod,
+  val payoutMethod: SelectedPaymentMethod,
+  val claims: List<String>
+) : MessageData
+
+class SelectedPaymentMethod(
+  val kind: String,
+  val paymentDetails: Map<String, Any>
```
Comment (ethan-tbd) [inline protocol/src/main/kotlin/models/MessageData.kt:20]: is `paymentDetails` in `SelectedPaymentMethod` meant to be required? The [tbdex protocol](https://github.com/TBD54566975/tbdex/tree/main/specs/protocol#selectedpaymentmethod) has it as not required but [tbdex-js](https://github.com/TBD54566975/tbdex-js/blob/9cf74c41eaff824e1746dcd40e9c544d1f8d632f/packages/protocol/src/types.ts#L127) seems to have it as required?

File: protocol/src/main/kotlin/Mapper.kt (hunk 4 above)
Comment (jiyoontbd) [reply protocol/src/main/kotlin/Mapper.kt:10]: I think it's just what we've been using. Is there an advantage to using kotlinx instead?

**Review by jiyoontbd — APPROVED** [review]
super strong 💪

File: protocol/src/main/kotlin/models/Message.kt (hunk 3 above)
Comment (andresuribe87) [reply protocol/src/main/kotlin/models/Message.kt:26]: Thanks @ethan-tbd - is that the right place long term for it?

File: protocol/src/main/kotlin/Mapper.kt (hunk 4 above)
Comment (andresuribe87) [reply protocol/src/main/kotlin/Mapper.kt:10]: The two advantages I know of:
* Speed
* Verbosity
I don't know how well it's supported in different versions of kotlin though, or different platforms.

Comment (KendallWeihe) [reply protocol/src/main/kotlin/Mapper.kt:10]: Relevant comment https://github.com/TBD54566975/tbdex-kt/pull/31#issuecomment-1747426398

**Review by ethan-tbd — APPROVED** [review]

File: protocol/src/main/kotlin/models/Message.kt [hunk 7]
```diff
@@ -1,89 +1,83 @@
...
-    }
+  // TODO - use web5 crypto and fix the types
+  fun sign(privateKey: String, kid: String) {
+    this.signature = "blah"
+  }
 
-    override fun toString(): String {
-        return Mapper.writer().writeValueAsString(this)
-    }
+  fun toJsonString(): String {
```
Comment (KendallWeihe) [inline protocol/src/main/kotlin/models/Message.kt:57]: I think the pattern is just...

```suggestion
  fun toJson(): String {
```

Which implies it's a string. I think?

File: protocol/src/main/kotlin/Mapper.kt (hunk 4 above)
Comment (phoebe-lew) [reply protocol/src/main/kotlin/Mapper.kt]: @andresuribe87 , this was a grandfathered decision from @bradleydwyer and @tomdaffurn. I personally wanted to be able to use kotlinx, but also heard from @alecthomas that it wasn't ready for prime time. Kotlinx has a MUCH nicer interface.

File: protocol/src/main/kotlin/models/MessageData.kt (hunk 6 above)
Comment (phoebe-lew) [reply protocol/src/main/kotlin/models/MessageData.kt:20]: Hmmmm I think it is not required...for example one method in the future might be a previously topped up stored balance, in which case you don't have any additional payment details to submit

File: protocol/src/main/kotlin/models/Message.kt (hunk 7 above)
Comment (phoebe-lew) [reply protocol/src/main/kotlin/models/Message.kt:57]: which pattern?

File: protocol/src/main/kotlin/models/Close.kt (hunk 2 above)
Comment (phoebe-lew) [reply protocol/src/main/kotlin/models/Close.kt:3]: I would suggest NOT rolling our own token lib and implementing typeid in languages where it's currently not supported.

File: protocol/src/test/kotlin/models/MessageTest.kt (hunk 1 above)
Comment (phoebe-lew) [reply protocol/src/test/kotlin/models/MessageTest.kt:1]: Didn't you have a PR open to fix that?

File: protocol/src/main/kotlin/models/MessageData.kt [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  val expiresAt: OffsetDateTime,
+  val payin: QuoteDetails,
+  val payout: QuoteDetails,
+  val paymentInstructions: PaymentInstructions? = null
+) : MessageData
+
+class QuoteDetails(
+  val currencyCode: String,
+  val amountSubunits: Int,
+  val feeSubunits: Int?
```
[the remark this answers is not in the document]
Comment (phoebe-lew) [reply protocol/src/main/kotlin/models/MessageData.kt:34]: yeah good call! makes sense for most of them

File: protocol/src/main/kotlin/models/Message.kt (hunk 7 above)
Comment (phoebe-lew) [reply protocol/src/main/kotlin/models/Message.kt:57]: omg sorry i'm dumb yes this makes sense lmao

File: protocol/src/main/kotlin/Mapper.kt (hunk 4 above)
Comment (alecthomas) [reply protocol/src/main/kotlin/Mapper.kt]: Yeah, I tried and failed to use kotlinx serialisation. I really liked the API too, but IIRC it was just generally quite restrictive and fiddly to use.

**Pull request merged**